crypto/encode_decode/encoder_lib.c

```diff
@@ -14,13 +14,41 @@
 #include <openssl/buffer.h>
 #include <openssl/params.h>
 #include <openssl/provider.h>
+#include <openssl/trace.h>
 #include "encoder_local.h"
 
-static int encoder_process(OSSL_ENCODER_CTX *ctx, BIO *out);
+struct encoder_process_data_st {
+    OSSL_ENCODER_CTX *ctx;
+
+    /* Current BIO */
+    BIO *bio;
+
+    /* Index of the current encoder instance to be processed */
+    int current_encoder_inst_index;
+
+    /* Processing data passed down through recursion */
+    int level;                   /* Recursion level */
+    OSSL_ENCODER_INSTANCE *next_encoder_inst;
+    int count_output_structure;
+
+    /* Processing data passed up through recursion */
+    OSSL_ENCODER_INSTANCE *prev_encoder_inst;
+    unsigned char *running_output;
+    size_t running_output_length;
+};
+
+static int encoder_process(struct encoder_process_data_st *data);
 
 int OSSL_ENCODER_to_bio(OSSL_ENCODER_CTX *ctx, BIO *out)
 {
-    return encoder_process(ctx, out);
+    struct encoder_process_data_st data;
+
+    memset(&data, 0, sizeof(data));
+    data.ctx = ctx;
+    data.bio = out;
+    data.current_encoder_inst_index = OSSL_ENCODER_CTX_get_num_encoders(ctx);
```

Comment:
Is there a reason for gong backwards?

Reply:
It seemed more deterministic that way...  but, maybe that angle disappears if I find a fix for what you commented in https://github.com/openssl/openssl/pull/13167#discussion_r516435139

Reply:
Nope, still seems simpler to go backwards.  Basically, we know that we want to end up with a specific output format, and potentially a specific outermost structure (such as PKCS#8).  That constitutes a deterministic and unique starting point.  Going from the other end leads to a more complex decision tree, as far as I can see (I admit, though, that I haven't experimented much with that, so this is more gut feeling than actual research.  feel free to research and correct me :wink:)